libs/zk/include/nil/crypto3/zk/commitments/detail/polynomial/basic_fri.hpp

```diff
@@ -774,155 +744,272 @@ namespace nil {
                             }
                         }
                     }
+                    return std::move(g_coeffs);
+                }
 
-                    for (std::size_t query_id = 0; query_id < fri_params.lambda; query_id++) {
-                        std::size_t domain_size = fri_params.D[0]->size();
-                        typename FRI::field_type::value_type x = transcript.template challenge<typename FRI::field_type>();
-                        x = x.pow((FRI::field_type::modulus - 1)/domain_size);
-                        std::uint64_t x_index = 0;
-                        for( x_index = 0; x_index < domain_size; x_index++ ){
-                            if( fri_params.D[0]->get_domain_element(x_index) == x ){
-                                break;
-                            }
-                        }
-                        t = 0;
-
-                        std::vector<std::array<typename FRI::field_type::value_type, FRI::m>> s;
-                        std::vector<std::array<std::size_t, FRI::m>> s_indices;
-                        std::tie(s, s_indices) = calculate_s<FRI>(x, x_index, fri_params.step_list[0], fri_params.D[0]);
 
-                        //Initial proof
-                        std::map<std::size_t, typename FRI::initial_proof_type> initial_proof;
-                        for (const auto &it: g) {
-                            auto k = it.first;
-                            initial_proof[k] = {};
-                            initial_proof[k].values.resize(it.second.size());
-                            std::size_t coset_size = 1 << fri_params.step_list[0];
-                            BOOST_ASSERT(coset_size / FRI::m == s.size());
-                            BOOST_ASSERT(coset_size / FRI::m == s_indices.size());
+                template<typename FRI, typename PolynomialType>
+                static std::map<std::size_t, typename FRI::initial_proof_type>
+                build_initial_proof(
+                    std::map<std::size_t, typename FRI::precommitment_type> const& precommitments,
```

Comment:
Let's move 'const' to the start please, I'm used to that:

const std::map<std::size_t, typename FRI::precommitment_type> &